Suggesting some changes on the Processor API update

smaldini opened the pull request:
Giving a go at my latest comment on #2188

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -203,64 +159,43 @@ public void error(Throwable e) {
 
-	//TODO improve synchronization, prefer CAS ?
-	static final class MonoProcessorSink<T> implements StandaloneMonoSink<T> {
+	static final class MonoProcessorSink<T> implements Sink<T> {
```
if sinks are advertised as thread safe objects, doesn't this require external synchronization of the processor's calls ? especially if we later introduce several flavors of `MonoProcessor` ?

smaldini replied:
MonoProcessor are safe, all the onXxx methods are atomic - I'll actually have an extra commit that moves further and remove the need for MonoProcessorSink by making MonoProcessor implementing it directly.

simonbasle replied:
ok. the later might be a better solution actually, since then there is no risk of adding a (non-serialized) flavor of `MonoProcessor` and mistakenly wrapping it with a (non-serialized) sink decorator

simonbasle commented on docs/asciidoc/processors.adoc:
```diff
@@ -45,7 +45,7 @@ If, after exploring the above alternatives, you still think you need a `Processo
 
 [[sinks]]
-= Safely Produce from Multiple Threads by Using `StandaloneFluxSink` and `StandaloneMonoSink`
+= Safely Produce from Multiple Threads by Using `Sink` and `Sink`
```
IF the mono flavor is removed, this `and 'Sink'` can be removed. but I'm not a huge fan of removing the mono flavor...

smaldini replied:
I think that is something I'm not too attached to personally if you think having a dedicated mono is valuable. I'll clean the doc in the final pass. The main problem I'm attacking is removing Processors which is more impacting IMO.

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/DirectProcessor.java:
```diff
@@ -81,7 +81,7 @@
  */
 @Deprecated
-public final class DirectProcessor<T> extends FluxIdentityProcessor<T> {
+public final class DirectProcessor<T> extends FluxProcessor<T, T> {
```
could you maybe revert the `FluxIdentityProcessor` removal for now, and offer it in a separate PR ? this distracts from the rest of the PR IMHO, as maybe 90% of these files are only changed due to this removal.

smaldini replied:
The PR is not finished and there will be more impact due to removing Processors and use of Processor directly wherever possible.

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/FluxIdentityProcessor.java:
```diff
@@ -1,26 +0,0 @@
- * 
- * @author Simon Baslé
- */
-public abstract class FluxIdentityProcessor<T> extends FluxProcessor<T, T> {
```
the rationale for adding this was that `FluxProcessor<String, String>` feels clunky, and was such a pain to migrate when removing deprecation warnings in the test code. Maybe that pain will be a good motivator for people to move to `Sink`? But maybe on the contrary it will just make them angry 😝

smaldini replied:
Yeah the rationale is to not use processors api at all in the future so no need to spend any more surface api for them.

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/FluxSink.java:
```diff
@@ -113,6 +113,16 @@
 
+	@Override
+	default Flux<T> toFlux() {
+		return Flux.error(new IllegalStateException("A FluxSink does not support back referencing the outer Flux"));
```
honestly, I'm not a huge fan of offering an API and then having it error in half the cases. this feels like a step backward towards the same kind of inconsistencies that we had in processors.

smaldini replied:
It's not impossible to support it but its easy to make mistakes if supported. At least its explicit that it doesn't work unlike using Processor directly vs sinks. We can look at the final result once I'm done and see what can go back. Right now i'm aiming for one API entry: Sink/Sinks and will add more from this.

simonbasle replied:
fixed in 6f60e9f

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/MonoSink.java:
```diff
@@ -31,7 +33,45 @@
  * @param <T> the value type emitted
  */
-public interface MonoSink<T> extends ScalarSink<T> {
+public interface MonoSink<T> extends Sink<T> {
```
I think this is very problematic. Offering a flux-like API on the `MonoSink`, even if we introduce a specialized API, is bound to be confusing.

ie this API seemingly allows `next(t).next(t).complete()` or even more confusing, `next(t1).next(t2).success(t3)` type of calls.

smaldini replied:
That is changed in the next commit. Sink has orthogonal methods "boolean emitXxx" (complete|error|next). Chaining only brings value to FluxSink if any and it was impossible for processors to be internally sink due to competing error name. Plus it provides for a new feature, returning false if emission failed because the sink is terminated.

simonbasle replied:
wouldn't it make implementing the `tryNext` feature harder ? I'm guessing this overlaps a bit, but it could be interesting to be able to distinguish between a `emitNext` call that returns `false` because termination from one that returns `false` because backpressure. wdyt @dfeist ?

smaldini replied:
If we go that route we always bring back the enum `Emission` with various helper `isOk`, `isTerminated`, `isCancelled`, `isDisposed`, `isBackpressured`

smaldini commented:
I'm close to happiness with the changes: 
- Only 2 new APIs : Sinks and Sink
- Remove the need to use a Processor altogether except for integration purpose
- Optimized Sink impls to avoid extra wrapper allocation

Follow up: refining Sinks/Sink api if needed:
- the alloc-free builder API is a good idea to make `Sinks` more guided, provide toProcessor 
- do we need extra Sink api: scannable(), isTerminated/hasCancelled, hasDowntream/Subscribers...
- polish doc

smaldini commented:
Regarding the builder suggestion here is the the proposal:
```java
		Sink<String> test = Sinks.multicast()
								 .onBackpressureBuffer()
								 .get();

		Sink<String> test2 = Sinks.replay()
								  .limit(Duration.ofMillis(100))
								  .get();

		Sink<String> test3 = Sinks.replay()
								  .limit(Duration.ofMillis(100))
								  .get();

		FluxProcessor<String, String> test4 = Sinks.unicast()
												   .onBackpressureBuffer()
												   .fluxProcessor();
```

simonbasle commented on reactor-core/src/test/java/reactor/core/publisher/FluxAutoConnectTest.java:
```diff
@@ -43,42 +43,42 @@ public void constructors() {
 	@Test
 	public void connectImmediately() {
-		FluxProcessor<Integer, Integer> e = Processors.multicast();
+		Sink<Integer> e = Sinks.multicast();
```
since the sink isn't really used, can also be replaced with `TestPublisher` which makes it easier to assert subscribers:

```java
		TestPublisher<Integer> testPublisher = TestPublisher.create();

		testPublisher.flux().publish().autoConnect(0, cancel::set);
		testPublisher.assertSubscribers();
		
		cancel.get().dispose();
		testPublisher.assertNoSubscribers();
```

simonbasle commented on reactor-core/src/test/java/reactor/guide/GuideTests.java:
```diff
@@ -201,21 +201,20 @@ public void advancedCold() {
 	@Test
 	public void advancedHot() {
-		Sink<String> hotSource = Sinks.multicastNoWarmup();
+		Sink<String> hotSource = Sinks.replay(0);
```
let's use `Sinks.multicast()`, `replay(0)` being kind of a corner case.
Also change the corresponding snippet in the asciidoc!

simonbasle commented on docs/asciidoc/advancedFeatures.adoc:
```diff
@@ -193,7 +193,7 @@ Compare the first example to the second example, shown in the following code:
 [source,java]
 ----
-Sink<String> hotSource = Sinks.multicastNoWarmup();
+Sink<String> hotSource = Sinks.replay(0);
```
see comment on `GuideTest.java`

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -47,25 +53,37 @@ private Sinks() { }
+
+	/**
+	 * A {@link Sink} with the following characteristics:
+	 * <ul>
```
Add a note that it is better suited to be viewed as a `toMono()` ?

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -47,25 +53,37 @@ private Sinks() { }
 	@SuppressWarnings("deprecation")
-	public static <T> Sink<T> multicastNoWarmup() {
-		return new ProcessorSink<>(ReplayProcessor.create(0));
+	public static Sink<Void> coordinator() {
```
put it next to `promise`, as both behave closer to `Mono`. That makes me wonder if it would be better backed by a `MonoProcessor`? likely to be used as `toMono()` more often and thus benefit from the 0-overhead conversion? or maybe add a constructor parameter to `DirectProcessor` that allows its `toMono()` method to do `Mono.fromDirect(this)` rather than `next()`?

smaldini replied:
In a separate PR we will have VoidProcessor for this

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -112,92 +132,73 @@ private Sinks() { }
 	 * A {@link Sink} that works like a conceptual Promise: it can be completed
 	 * with or without a value at any time, but only once. This completion is replayed to late subscribers.
-	 * Calling {@link Sink#next(Object)} is enough and does not require a following {@link Sink#complete()}.
+	 * Calling {@link Sink#emitNext(Object)} is enough and does not require a following {@link Sink#emitComplete()}.
```
add ` (which is implicit).` at the end

simonbasle replied:
also add that as a result the `Sinks` closely follows the `Mono` contract and is better suited to be viewed as `toMono()` (hinting at 0-overhead conversion to mono vs overhead conversion to flux)

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -112,92 +132,73 @@ private Sinks() { }
-			return this;
-		}
+	/**
+	 * Return unsafe {@link Sink} that are not serialized and expect the signal producer to respect
```
change to "and expect usage to be externally synchronized so as to respect"

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/EmitterProcessor.java:
```diff
@@ -192,6 +192,85 @@ public void subscribe(CoreSubscriber<? super T> actual) {
+		}
+
+		while (!q.offer(t)) {
+			LockSupport.parkNanos(10);
```
I think we should just not support it at all this is a mistake and it feels like a documentable change for 3.4

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/FluxCreate.java:
```diff
@@ -131,102 +130,102 @@ public Object scanUnsafe(Attr key) {
 			for (; ; ) {
 
 				for (; ; ) {
-					if (e.isCancelled()) {
```
double checked but unfortunately there is no easy other way for now and i don't want to scope creep into changing this here. So reverted

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/FluxGroupJoin.java:
```diff
@@ -207,7 +207,7 @@
 		public Stream<? extends Scannable> inners() {
 			return Stream.concat(
-					lefts.values().stream(),
+					lefts.values().stream().map(Scannable.class::cast),
```
I'm not too sure, tangent, inners should have its own attribute (and delegate to it on the Scannable contract).
Will update in another PR if we want

smaldini replied:
good catch

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/FluxSink.java:
```diff
@@ -33,9 +34,29 @@
 
-	@Override
+	/**
+	 * Emit a non-null element, generating an {@link Subscriber#onNext(Object) onNext} signal.
```
I'd tend to agree, you still have some chaining value out of next but error and complete are just there to align semantically with next. To be reviewed in a subsequent PR imo

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/MonoSink.java:
```diff
@@ -55,11 +55,18 @@
+	void error(Throwable e);
+
+	/**
+	 * Terminate with the given value without requiring {@link #emitComplete()} to be explicitly called.
```
Agree - Being consistent here with the hook there and processor/sink. That might be a separate PR to Improve documentation on Dropping behaviors.

smaldini commented:
Besides rebasing, the last touch on this PR : 
- If the guided Sinks api works for everyone, should we extend it to promise, adding its doc too : 
```java
Sink<Void> coordinator = Sinks.promise().coordinator(); // I have implemented a `VoidProcessor extends MonoProcessor` locally

Sink<String> nextPromise = Sinks.promise().next(); //the current MonoProcessor, locally moved under NextProcessor.

Sink<String> latestPromise = Sinks.promise().latest(); // a new processor I wanted to suggest that would emit the latest value only, useful for one off observers.

```
- Adding some common state observers instead of relying Scannable which is not a guaranteed state query: hasSubscribers, isTerminated(), hasError() / getError()
- For this pr or later changing emitterprocessor lock.parkNanos to onError: I don't think we should implement any sort of backpressure management on Sink, also i don't know about adding "offer()" or similar: Flux.create can already do that job in a more appropriate way using onRequest.

In a later iteration I'd like to see a `toConnectableMono` and `toConnectableFlux` on the sink while deprecating autoCancel on emitterProcessor. There should be a common way to address disposing between all sinks that is similar to Flux/Mono : refCount, connect etc. By default all `Sinks` except EmitterProcessor are similar to `share()`

simonbasle commented on reactor-core/src/test/java/reactor/core/publisher/UnicastProcessorTest.java:
```diff
@@ -65,20 +65,19 @@ public void secondSubscriberRejectedProperly() {
 	public void multiThreadedProducer() {
-		UnicastProcessor<Integer> processor = UnicastProcessor.create();
-		FluxSink<Integer> sink = processor.sink();
+		Sink<Integer> sink = Sinks.unicast().onBackpressureBuffer();
```
I'm not sure `xxxProcessor` tests should use the `Sinks` path. After all, these classes are there to explicitly test a specific concrete implementation and would likely survive the switch of said concrete implementation to package-private. On the other hand, if the implementation of `Sinks.unicast()` somehow changes then coverage would drop for `UnicastProcessor`.

This comment applies to other `xxxProcessorTest` cases.

simonbasle commented on docs/asciidoc/advancedFeatures.adoc:
```diff
@@ -193,7 +193,7 @@ Compare the first example to the second example, shown in the following code:
 [source,java]
 ----
-Sinks.StandaloneFluxSink<String> hotSource = Sinks.multicastNoWarmup();
+Sink<String> hotSource = Sinks.replay().limit(0);
```
see comment in `GuideTest.java`

simonbasle replied:
fixed in 7b6b655

simonbasle commented on reactor-core/src/test/java/reactor/guide/GuideTests.java:
```diff
@@ -201,21 +201,20 @@ public void advancedCold() {
-		Sinks.StandaloneFluxSink<String> hotSource = Sinks.multicastNoWarmup();
-
-		Flux<String> hotFlux = hotSource.asFlux().map(String::toUpperCase);
+		Sink<String> hotSource = Sinks.replay().limit(0);
```
reiterating that `replay().limit(0)` is probably a bit too confusing. `multicast().onBackpressureError()` sounds like the best example to give, showcasing only live data makes it past the sink.

Also change in `advancedFeatures.adoc`.

simonbasle replied:
fixed in 7b6b655

simonbasle commented:
> ```java
> Sink<Void> coordinator = Sinks.promise().coordinator(); // I have implemented a `VoidProcessor extends MonoProcessor` locally
> 
> Sink<String> nextPromise = Sinks.promise().next(); //the current MonoProcessor, locally moved under NextProcessor.
> 
> Sink<String> latestPromise = Sinks.promise().latest(); // a new processor I wanted to suggest that would emit the latest value only, useful for one off observers.
> ```

I'm good with introducing a spec for mono processors as a follow up, but I think these changes *should be done in a separate PR*. `VoidProcessor` sounds good by itself, but the rework of `MonoProcessor` makes me feel this doesn't belong in this particular PR. Plus there are two dedicated issues for that and `LatestProcessor` : #1053 and #1310 (looks like the second one could actually cover both `next()` and `latest()`). On a side note the names for the API would need some more thought.


> 
>     * Adding some common state observers instead of relying Scannable which is not a guaranteed state query: hasSubscribers, isTerminated(), hasError() / getError()

Ok yeah, why not.

> 
>     * For this pr or later changing emitterprocessor lock.parkNanos to onError: I don't think we should implement any sort of backpressure management on Sink, also i don't know about adding "offer()" or similar: Flux.create can already do that job in a more appropriate way using onRequest.

Since there is a dedicated issue (#2049) it would make sense in another PR. Would greatly appreciate if you have the time to tackle it 🙇 
> 
> 
> In a later iteration I'd like to see a `toConnectableMono` and `toConnectableFlux` on the sink while deprecating autoCancel on emitterProcessor. There should be a common way to address disposing between all sinks that is similar to Flux/Mono : refCount, connect etc. By default all `Sinks` except EmitterProcessor are similar to `share()`

Yeah that sounds like a nice improvement in consistency. The only thing is we don't have `ConnectableMono` :)

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/DelegateSinkFluxProcessor.java:
```diff
@@ -0,0 +1,120 @@
+
+	@Override
+	public Context currentContext() {
+		if(upstream instanceof CoreSubscriber){
```
The same existed in FluxProcessor but yeah upstream is the receiving side and downstream the producing side here. Can change to less ambiguous term tho

simonbasle replied:
yeah, had to double-back during the review. I think it would benefit readability for our future selves 😄

rstoyanchev commented:
I like where this has gone from a high-level usage perspective, it looks much cleaner.

One suggestion for consideration. The terminology for the entry points in `Sinks` is a bit mixed. It's not as self-evident as could be. For example `coordinator` and `promise` stand out the most. We probably don't want to introduce such terms that compete with `Mono<T>` and `Mono<Void>`. For such method names it probably won't be too hard to find something more plain.

One thought I had it is something like this:

- `Sinks`
  - `completion()` -> `Sink<Void>`
  - `item()` -> `Sink<T>`
  - `sequence()` -> `SequenceSpec`
    - `unicast()` -> `UnicastSpec`
    - `multicast()` -> `MulticastSpec`
    - `replay()` -> `ReplaySpec`

smaldini commented:
Hey @rstoyanchev and @simonbasle I'll get back to this over this weekend unless i can push some update today. Locally I stopped on having to separate the work on MonoProcessor to support Mono<Void> in a different PR. 
I agree with the suggestion you made @rstoyanchev in terms of organizing first by cardinality. In fact I wondered about : 
```Sinks.zero(), Sinks.one(), Sinks.many()``` 

The reason is to deemphasize on the item (`Sinks.one` could be at most one where `item()` feels like you need to produce one item). Variants include `Sinks.empty()`, `Sinks.singleOrEmpty()` and `Sinks.many()` but we can adapt naming at will once the concept is in. 
So leftover is to polish with @simonbasle comments and move the rest to a further iteration (scannable and connectable additions).

smaldini commented:
Better late than never they said - So in this push :
- Split the PR (the other one on MonoProcessor to come)
- The reorg using Sinks.empty, singleOrEmpty and many
- Added the Emission enum.
- Also changed the EmitterProcessor behavior to not lock

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/SinksSpecs.java:
```diff
@@ -0,0 +1,174 @@
+
+}
+
+final class FluxProcessorSink<T> extends FluxCreate.SerializedSink<T, Sink<T>> {
```
splitted out and renamed in 6f60e9f

simonbasle commented:
It should be noted that this PR incidentally fixes #2049

simonbasle commented on reactor-core/src/main/java/reactor/core/publisher/Sinks.java:
```diff
@@ -87,118 +78,315 @@ private Sinks() { }
+		 *     <li>Without {@link Subscriber}: elements pushed via {@link Sink#emitNext(Object)} are discarded</li>
+		 *     <li>Backpressure : this sink is not able to honor downstream demand and will emit `onError` if there is a mismatch.</li>
+		 *     <li>Replaying: No replay. Only forwards to a {@link Subscriber} the elements that have been
+		 *     pushed to the sink AFTER this subscriber was subscribed. To the exception of the first
```
fixed in 8cb64857c

simonbasle commented:
@reactor/core-team I have reworked the PR according to our internal discussions, can you review and approve?

aneveu commented on reactor-core/src/test/java/reactor/core/publisher/FluxFilterWhenTest.java:
```diff
@@ -254,13 +254,13 @@ protected void hookFinally(SignalType type) {
 	@Test
 	public void cancel() {
-		final FluxIdentityProcessor<Boolean> pp = Processors.multicast();
+		final Sinks.Many<Boolean> pp = Sinks.many().multicast().onBackpressureBuffer();
```
not really important but the name of the variable was already quite unclear before and it becomes completely impossible to link to anything now 😅
